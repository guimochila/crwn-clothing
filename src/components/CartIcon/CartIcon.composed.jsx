import React from 'react';
import { Mutation, Query } from 'react-apollo';
import { adopt } from 'react-adopt';

import CartIcon from './CartIcon';
import { TOGGLE_CART_HIDDEN, GET_ITEM_COUNT } from '../../graphql/queries';

const Composed = adopt({
  toggleCartHidden: ({ render }) => (
    <Mutation mutation={TOGGLE_CART_HIDDEN}>{render}</Mutation>
  ),
  localState: ({ render }) => <Query query={GET_ITEM_COUNT}>{render}</Query>,
});

const CartIconComposed = () => (
  <Composed>
    {({ toggleCartHidden, localState }) => (
      <CartIcon
        toggleCartHidden={toggleCartHidden}
        itemCount={localState.data.itemCount}
      />
    )}
  </Composed>
);

export default CartIconComposed;
