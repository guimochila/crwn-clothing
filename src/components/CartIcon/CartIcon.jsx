import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import {
  CartIconContainer,
  ShoppingIcon,
  ItemCountContainer,
} from './CartIcon.styles';
import { toggleCartHidden } from '../../store/cart/cart.actions';
import { selectCartItemsCount } from '../../store/cart/cart.selectors';

export function CartIcon({ toggleCartHidden, itemCount }) {
  return (
    <CartIconContainer onClick={toggleCartHidden}>
      <ShoppingIcon />
      <ItemCountContainer>{itemCount}</ItemCountContainer>
    </CartIconContainer>
  );
}

const mapStateToProps = createStructuredSelector({
  itemCount: selectCartItemsCount,
});

const mapDispatchToProps = dispatch => ({
  toggleCartHidden: () => dispatch(toggleCartHidden()),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(CartIcon);
