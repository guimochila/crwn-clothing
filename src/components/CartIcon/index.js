import CartIconComposed from './CartIcon.composed';

export default CartIconComposed;
