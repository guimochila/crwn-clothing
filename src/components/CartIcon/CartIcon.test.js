import React from 'react';
import { shallow } from 'enzyme';

import { CartIcon } from './CartIcon';

describe('CartIcon component', () => {
  let wrapper;
  let mockToggleCartHidden;

  beforeEach(() => {
    mockToggleCartHidden = jest.fn();

    const mockProps = {
      toggleCartHidden: mockToggleCartHidden,
      itemCount: -0,
    };

    wrapper = shallow(<CartIcon {...mockProps} />);
  });

  test('it should render CartIcon', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should call toggleCartHidden when clicked', () => {
    wrapper.find('CartIconContainer').simulate('click');
    expect(mockToggleCartHidden).toHaveBeenCalledTimes(1);
  });

  test('should render itemCount as a text', () => {
    const itemCount = parseInt(wrapper.find('ItemCountContainer').text());
    expect(itemCount).toBe(0);
  });
});
