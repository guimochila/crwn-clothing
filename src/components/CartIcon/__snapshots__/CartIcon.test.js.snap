// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartIcon component it should render CartIcon 1`] = `
<CartIconContainer
  onClick={[MockFunction]}
>
  <Styled(Component) />
  <ItemCountContainer>
    -0
  </ItemCountContainer>
</CartIconContainer>
`;
